// @graphql-codegen/cli/temp/output-graphql.tsx
export type Maybe<T> = T | null;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type MyType = {
  f?: Maybe<Scalars['String']>;
};

export type Query = {
  f?: Maybe<Scalars['String']>;
};
export type MyQueryQueryVariables = {};

export type MyQueryQuery = { __typename?: 'Query' } & Pick<Query, 'f'>;

import gql from 'graphql-tag';
import * as React from 'react';
import * as ReactApollo from 'react-apollo';
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export const MyQueryDocument = gql`
  query MyQuery {
    f
  }
`;

export const MyQueryComponent = (props: Omit<Omit<ReactApollo.QueryProps<MyQueryQuery, MyQueryQueryVariables>, 'query'>, 'variables'> & { variables?: MyQueryQueryVariables }) => (
  <ReactApollo.Query<MyQueryQuery, MyQueryQueryVariables> query={MyQueryDocument} {...props} />
);

export type MyQueryProps<TChildProps = {}> = Partial<ReactApollo.DataProps<MyQueryQuery, MyQueryQueryVariables>> & TChildProps;
export function withMyQuery<TProps, TChildProps = {}>(operationOptions?: ReactApollo.OperationOption<TProps, MyQueryQuery, MyQueryQueryVariables, MyQueryProps<TChildProps>>) {
  return ReactApollo.withQuery<TProps, MyQueryQuery, MyQueryQueryVariables, MyQueryProps<TChildProps>>(MyQueryDocument, {
    alias: 'withMyQuery',
    ...operationOptions,
  });
}
