// @graphql-codegen/typescript/dist/esnext/visitor.js
import { transformComment, wrapWithSingleQuotes, DeclarationBlock, indent, BaseTypesVisitor } from '@graphql-codegen/visitor-plugin-common';
import * as autoBind from 'auto-bind';
import { Kind } from 'graphql';
import { TypeScriptOperationVariablesToObject } from './typescript-variables-to-object';
export class TsVisitor extends BaseTypesVisitor {
    constructor(schema, pluginConfig, additionalConfig = {}) {
        super(schema, pluginConfig, {
            avoidOptionals: pluginConfig.avoidOptionals || false,
            maybeValue: pluginConfig.maybeValue || 'T | null',
            constEnums: pluginConfig.constEnums || false,
            enumsAsTypes: pluginConfig.enumsAsTypes || false,
            immutableTypes: pluginConfig.immutableTypes || false,
            ...(additionalConfig || {}),
        });
        autoBind(this);
        this.setArgumentsTransformer(new TypeScriptOperationVariablesToObject(this.scalars, this.convertName, this.config.avoidOptionals, this.config.immutableTypes));
        this.setDeclarationBlockConfig({
            enumNameValueSeparator: ' =',
        });
    }
    clearOptional(str) {
        if (str.startsWith('Maybe')) {
            return str.replace(/Maybe<(.*?)>$/, '$1');
        }
        return str;
    }
    NamedType(node) {
        return `Maybe<${super.NamedType(node)}>`;
    }
    ListType(node) {
        return `Maybe<${super.ListType(node)}>`;
    }
    wrapWithListType(str) {
        return `${this.config.immutableTypes ? 'ReadonlyArray' : 'Array'}<${str}>`;
    }
    NonNullType(node) {
        const baseValue = super.NonNullType(node);
        return this.clearOptional(baseValue);
    }
    FieldDefinition(node, key, parent) {
        const typeString = node.type;
        const originalFieldNode = parent[key];
        const addOptionalSign = !this.config.avoidOptionals && originalFieldNode.type.kind !== Kind.NON_NULL_TYPE;
        const comment = transformComment(node.description, 1);
        return comment + indent(`${this.config.immutableTypes ? 'readonly ' : ''}${node.name}${addOptionalSign ? '?' : ''}: ${typeString},`);
    }
    InputValueDefinition(node, key, parent) {
        const originalFieldNode = parent[key];
        const addOptionalSign = !this.config.avoidOptionals && originalFieldNode.type.kind !== Kind.NON_NULL_TYPE;
        const comment = transformComment(node.description, 1);
        return comment + indent(`${this.config.immutableTypes ? 'readonly ' : ''}${node.name}${addOptionalSign ? '?' : ''}: ${node.type},`);
    }
    EnumTypeDefinition(node) {
        const enumName = node.name;
        // In case of mapped external enum string
        if (this.config.enumValues[enumName] && typeof this.config.enumValues[enumName] === 'string') {
            return null;
        }
        if (this.config.enumsAsTypes) {
            return new DeclarationBlock(this._declarationBlockConfig)
                .export()
                .asKind('type')
                .withComment(node.description)
                .withName(this.convertName(node))
                .withContent('\n' +
                node.values
                    .map(enumOption => {
                    let enumValue = enumOption.name;
                    const comment = transformComment(enumOption.description, 1);
                    if (this.config.enumValues[enumName] && typeof this.config.enumValues[enumName] === 'object' && this.config.enumValues[enumName][enumValue]) {
                        enumValue = this.config.enumValues[enumName][enumValue];
                    }
                    return comment + indent(wrapWithSingleQuotes(enumValue));
                })
                    .join(' |\n')).string;
        }
        else {
            return new DeclarationBlock(this._declarationBlockConfig)
                .export()
                .asKind(this.config.constEnums ? 'const enum' : 'enum')
                .withName(this.convertName(node))
                .withComment(node.description)
                .withBlock(this.buildEnumValuesBlock(enumName, node.values)).string;
        }
    }
}
